TheJumpCloud/support
# SA-3114: Ad Hoc User Cert Distribution

## Issues
[SA-3114](https://jumpcloud.atlassian.net/browse/SA-3114) - Ad Hoc User Cert Distribution

## What does this solve?
With this latest version of the Radius Cert Distribution Scripts, we've made a few changes that should make the overall process more cohesive and streamlined.  The main points of this latest release are as follows:

- Ad Hoc user cert generation and distribution
- Interactive Menu
- Updated monitoring reports

---
### Ad Hoc User Cert Generation and Distribution
This will address if a user is added to the RADIUS user group after an admin has already generated and distributed certs to the existing user group members. 

https://user-images.githubusercontent.com/89030113/217340418-0dc843a2-3b94-4b86-997a-92a0dffa9a02.mov

We have also reduced the bloat of the previous iteration that created a unique command for each user->system association. Now there will be a maximum of 2 commands per user depending on their system associations. As an example:

- **John Doe** is assigned to the following workstations:
    - `JC-Windows10-01`
    - `JC-Windows10-02`
- **Bob Ross** is assigned to the following workstations:
    - `JC-Windows10-03`
    - `JC-MacOSX-01`

**John Doe** will only have 1 command created with the following naming structure:
- `RadiusCert-Install:John.Doe:Windows`
    - That command will then be associated to both `JC-Windows10-01` and `JC-Windows10-02`

**Bob Ross** will have 2 commands created with the following naming structure:
- `RadiusCert-Install:Bob.Ross:Windows`
    - That command will be associated to `JC-Windows10-03`
- `RadiusCert-Install:Bob.Ross:MacOSX`
    - That command will be associated to `JC-MacOSX-01`

### Interactive Menu
With the addition of various Interactive Menus, this helps reduce confusion for the admin as to what the next step in the process is. It will eliminate the need for the admin to manually call individual scripts and replaces it with prompts for the admin to select

![image](https://user-images.githubusercontent.com/89030113/217339079-a044dafd-7c80-4b4b-8558-a6acb25d69ea.png)

### Updated Monitoring Reports
We have updated the monitoring reports to be more dynamic and allow admins to see an overview report that contains the command name, the system it was assigned to and the status (OK, PENDING, FAILED). We have also created a failed report, which will give a more detailed output regarding which command failed on which system with the command result's output on the failure.

![image](https://user-images.githubusercontent.com/89030113/217339220-80d83857-a0da-459a-ace1-249ba1e00ef7.png)

![image](https://user-images.githubusercontent.com/89030113/217339304-e6383d59-4e10-43f3-8d0f-5cf2f8c3c41a.png)

## Initial Run Video

https://user-images.githubusercontent.com/89030113/217351747-b8ee4db2-e6f6-419b-9d0c-39c100a9ca91.mov

## Is there anything particularly tricky?
There are a few things that should be noted with how the RADIUS Cert scripts function
- The scripts must be run from the `scripts/automation/Radius` folder
- In it's current state, it does not take into account users that are REMOVED from the user group
- In it's current state, when generating new commands it does not take into account new system associations for users
- If command results are deleted/removed the monitor reports may not be accurate

## How should this be tested?
1. **Initial run**:
    - Create a user group in JC Console and assign users
    - Assign users to various systems
    - Pull the branch and edit the Config.ps1 file
    ```
    # READ/ WRITE API KEY
    $JCAPIKEY = 'APIKEY'
    # JUMPCLOUD ORGID
    $JCORGID = 'ORGID'
    # JUMPCLOUD USER GROUP
    $JCUSERGROUP = 'USERGROUPID'
    # USER CERT PASSWORD (user must enter this when importing cert)
    $JCUSERCERTPASS = 'secret1234!'
    # USER CERT Validity Length (days)
    $JCUSERCERTVALIDITY = 90
    ```
    - cd to the `scripts/automation/Radius` folder
    - Run the following script `./Start-RadiusDeployment.ps1`
    - Follow prompts in order, (1, 2, 3, 4)
2. **Add new user to RADIUS user group**
    - Add a new user to the RADIUS user group
    - cd to the `scripts/automation/Radius` folder
    - Run the following script `./Start-RadiusDeployment.ps1`
    - Select option 2 to update user certs
        - A new cert should be generated for the newly assigned user
    - Select option 3 to distribute the user cert
    - Select option 2 to generate a new command for the newly assigned user
        - A new command should be created for that new user
    - When back at the main menu, select option 4 to monitor the results and validate that they are correct for your current environment configuration


[SA-3114]: https://jumpcloud.atlassian.net/browse/SA-3114?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
